Use keypad.Keys module from CP7.0 to support boards with key per GPIO

eltariel opened the pull request:
These changes are intended to support the [Pimoroni Keybow2040](https://shop.pimoroni.com/products/keybow-2040) (same as #193) - a 4x4 grid with an RGB LED per key. Rather than directly modifying `KMKKeyboard` to accept a different matrix, I subclassed it to `CustomMatrixKeyboard` and overrode everything that works directly with the `MatrixScanner` class.

The new scanner implements as little as possible - it's basically just a key mapping / event translation layer over the top of `keypad.Keys`. Implementing a proper matrix should be a fairly trivial variation on this, using `keypad.KeyMatrix` instead (I don't have HW to test with although I could throw something together basic pretty easily).

My next step from here probably would be to refactor `KMKKeyboard` to pull everything but the `MatrixScanner` init code out into a common base class that we can use for the existing matrix as well as the CP7.0 keypad stuff. An extension on this would be a scanner that combines other scanners, so you could mix and match key-per-gpio and a regular matrix on a single board.

Hell, we could even have a scanner that reads encoder events, or one that reads UART or i2c events, or... etc 😛 

(EDIT: this implements #237, despite it not being what I set out to do originally...)

xs5871 commented:
Sorry there hasn't been an update in a while.
I did some additional work:
- updates to the scanner base class
- preliminary work to integrate with split keyboards. I don't have the hardware on hand to test it and I currently don't have the capacity to build it. I'd be surprised if it would work ootb.
- implementation of an aggregate scanner and scratched the plain scanner list.

`coordinate_mapping` hurts my brain, because its somewhat backwards. As it is now, it maps virtual keys to hardware keys, that is virtual keys are the indices and hardware keys are the value -- but it's supposed to translate hardware keys to virtual ones. It's possible to map multiple virtual keys to the same hardware key, which makes conceptually no sense (although only the first one would be used). I wish it was the other way around,... but anyway.
That ties into `intifiy_coordinates`. I don't see a way to re-implement it to play nicely with continuous `key_number`s without passing it row and column count, and at that point something like `row*num_colums+column` is seems like the less convoluted approach. Personally it'd bite the bullet and break backwards compatibility at this point. I'm open for discourse.

@eltariel: how's it going on your end. I'd still rather collaborate than maintain my own fork, though I really don't want to carry over the matrix-centric bytearray report.

FoamyGuy commented:
@eltariel or @xs5871 Is there any further update on this effort? I am very interested in seeing the support for `keypad` finished up and merged in. 

If you are having less time to dedicate towards it now I may be able to pick it up and try to get it wrapped up, though there is still a fair amount of things I need to get my head wrapped around before I could really dive in to this specifically. I currently have somewhat little experience with KMK (set up a 2 key "hello world" type example to learn the basics). But am looking to get more involved and most of the hardware that would be easiest for me to tinker with while learning are individual IO pins per key rather than matrices

xs5871 commented:
There haven't been any updates; what's still missing (at least from my implementation):
* documentation
* test with new release of circuitpython, as there was a bug in the key-hashing at the time
* test with split keyboards (can't do that myself)

It's not that I don't have the time to finish this. I was hoping to hear from @eltarial again, as this is their PR and we didn't merge our implementations.

FoamyGuy commented:
@xs5871 Thank you. Can you link me to the branch your work is in? I assume here: https://github.com/xs5871/kmk_firmware/tree/circuitpython-7-keypad would be great to confirm though so I can be sure I'm starting from the right spot. 

I'd be willing to help with documentation.

I can also test with the newest release of CircuitPython

I can try to test with a split keyboard, but am a bit less confident in this one than the other two. I don't have a split keyboard ready to go that I can test with so I'd need to build something. But I may be able to make some sort of minimal split device on a breadboard for baseline testing.

We'll see what eltariel says. If it makes sense to merge together the work from the branch in this PR with yours we could do that. Or if not and they do not have the time to dedicate any more maybe we can make a new PR from your branch after the above gets resolved.

xs5871 commented:
@FoamyGuy that is the correct and up-to-date branch.

Checked with cp 7.1.0-beta.1 and now instead of always getting the same hash for different objects, we get different hashes every time, even if the events should compare equal according to the documentation... Oh, well.
That doesn't impact functionality though in any way whatsoever.

> I can try to test with a split keyboard, but am a bit less confident in this one than the other two. I don't have a split keyboard ready to go that I can test with so I'd need to build something. But I may be able to make some sort of minimal split device on a breadboard for baseline testing.

That would be awesome.
I would hold off on documentation until we decided which branch we're using.

> We'll see what eltariel says. If it makes sense to merge together the work from the branch in this PR with yours we could do that. Or if not and they do not have the time to dedicate any more maybe we can make a new PR from your branch after the above gets resolved.

Sure, let's do that. Looking forward to getting this issue on the way.

jpconstantineau commented:
@FoamyGuy I have lots of split keyboards and a few more getting ready...
Ping me when you are ready to test the split portion.  I can test wireless and wired. It would be great to be able to get both Serial and I2C working as some boards use one vs the other...

I'll be able to test with the nrf52840 and the rp2040.

jpconstantineau commented:
@FoamyGuy I also have splits with matrices, and other simple keys (Keys vs KeyMatrix).
Most splits are symmetrical (Let's Split and all derivatives). Some may not be (TKL with separate numpad).  Taking care of the symmetrical ones with encoders on both sides will take care of 95% of all cases.

eltariel commented:
Hey all, sorry for the radio silence on this. I've stalled a bit since my main dev machine died and I've only just resurrected it, and $work has been chaos as well so I haven't had a lot of time.

That said, I had a bunch of updates just about ready to go before the damn PC died. I'd cleaned up a bunch of things including killing off `coordinate_mapping` altogether in favour of letting the `AggregateScanner` I wrote merge it's scanners into a single linear keymap. It was working locally with a keybow2040 split into two virtual matrices, and I was going to try to make two of them co-operate over serial or i2c next. I also have a split board and a macro pad on the way from @jpconstantineau so I'll be able to test this a bit more thoroughly when it shows up, but... I have to dig it out from an old ssd first 😖

I've got some time off coming up soon, in theory I'll get some more done on this soon. I want to dig into @xs5871 's branch too, it sounds like we were converging on a similar solution and it's far more productive to collaborate than compete 😄

xs5871 commented:
@eltariel: any progress on your front? We have pending PR for the scanner interface to use the CP `KeyEvents`. Separating the Matrix code and allowing different scanners should be fairly trivial without breaking functionality.

eltariel commented:
Last couple of commits are ready to go. A few things are broken in the boards I've added, but that's unrelated to the keypad changes.

This PR has the following changes:
- Scanner classes
  - new `Scanner` base class based on the interface for the existing `Matrix`
  - `Matrix` implements `Scanner`
  - Added `NativeKeypadScanner` with factory methods to create scanners based on `keypad.Keys` and `keypad.KeyMatrix`
- Added check for `None` before creating the scanner in `KMKKeyboard`, so you can inject an alternate scanner
- Added `pixels` parameter to the `RGB` extension to allow using something other than Neopixels - works well so far with APA102 (see keybow for an example)
- Two new boards, both with key-per-GPIO and unusual RGB LEDs:
  - Pimoroni Keybow 2040: 4x4 with an RGB LED per key. Custom RP2040 board with an IS31FL3731 LED controller (currently LEDs aren't working).
  - Pimoroni Keybow: 4x3 with an APA102 RGB LED per key in a Raspberry Pi hat. Doesn't work properly on the rpi circuitpython port, but functions perfectly with an itsybitsy via an adapter board, I'm using it with an nRF52840, but should work with minimal changes for any of the itsybitsy boards.

eltariel commented:
Splitting this up into separate PRs for the scanner, the RGB changes, and the pimoroni boards.
